Setup linter for HTML

loicbourgois opened the pull request:
A comment from @benmccann made me realise the Javascript linter was not checking the files in /samples. [Comment is here](https://github.com/chartjs/Chart.js/pull/5134/files/a01f580f9905e28ecea3ad3393e64af7647527f3).

This is a PR to setup a linter for HTML files using [html plugin for eslint](https://github.com/BenoitZugmeyer/eslint-plugin-html) and [htmllint](https://github.com/yvanavermaet/gulp-htmllint).

The rules for `htmllint` can be modified in `.htmllintrc.js` . List of options is available [here](https://github.com/htmllint/htmllint/wiki/Options).

In `gulpfile.js` I added the option `lint-script-elements` to lint (or not) the HTML files in `lintTask()`.
Enabling this option gives 4166 errors. 
```
$ gulp lint --lint-script-elements
✖ 4230 problems (4166 errors, 64 warnings)
  3481 errors, 0 warnings potentially fixable with the `--fix` option.
```

What would be the best way to update the examples : file by file, folder by folder, all at once ?

loicbourgois commented:
Some more details about the errors : 
* 2606 `Expected indentation of ...`
* 603 `Strings must use singlequote`
* 598 `is not defined`
  * 499 `'randomScalingFactor' is not defined`
  * 80 `'Chart' is not defined`
  * 19 other
* 359 other

loicbourgois commented:
I think HTML files should at least be checked for indentation. It will be a lot easier to have consistency between the content of `<script>` elements and the rest of the DOM.

@simonbrunel I'm not sure how to discuss this on slack : I tried the badge on the readme, and was redirected to https://chart-js-automation.herokuapp.com/ with a message "There's nothing here, yet.".

simonbrunel commented:
> I think HTML files should at least be checked for indentation

Let's talk about that on Slack / another PR since I doubt ESLint can handle that.

> https://chart-js-automation.herokuapp.com/

Right, it's broken but I don't have access to fix it (@etimberg ?)

simonbrunel commented:
@tannerlinsley actually, that might be you that setup the heroku app?
@loicbourgois I sent you an invitation to join Slack

loicbourgois commented on samples/advanced/progress-bar.html:
```diff
@@ -2,95 +2,94 @@
-                }
-            }
-        };
+		<div style="width: 75%;">
```
need to find a rule to not allow 2 tabs

loicbourgois commented on samples/advanced/data-labelling.html:
```diff
@@ -3,130 +3,130 @@
+				// Define a plugin to provide data labels
+				Chart.plugins.register({
+					afterDatasetsDraw: function(chart/* , easing */) { // TODO : remove easing ?
+						// To only draw at the end of animation, check for easing === 1
```
comment and code would be out of sync

benmccann replied:
I would remove both the variable and the comment

loicbourgois commented on samples/charts/line/stepped.html:
```diff
@@ -29,14 +29,14 @@
 				labels: ['Day 1', 'Day 2', 'Day 3', 'Day 4', 'Day 5', 'Day 6'],
 				datasets: [{
-					label: 'steppedLine: ' + ((typeof(details.steppedLine) === 'boolean') ? details.steppedLine : `'${details.steppedLine}'`),
+					label: 'steppedLine: ' + ((typeof (details.steppedLine) === 'boolean') ? details.steppedLine : details.steppedLine), // TODO : check this
```
not sure what's going on here

simonbrunel replied:
`typeof` as operator without parenthesis?

loicbourgois replied:
no, i mean why we have 
``` 
? details.steppedLine : `'${details.steppedLine}'`

simonbrunel replied:
I think `typeof(foo)` is invalid because it's an operator so should not be used with parenthesis. We should write `typeof foo`, not `typeof (foo)` (with the extra space).

Anyway, you right, this line is weird, could be: `label: 'steppedLine: ' + details.steppedLine`

loicbourgois commented on samples/scales/time/line-point-data.html:
```diff
@@ -122,10 +122,13 @@
 		});
 
-		document.getElementById('addData').addEventListener('click', function() {
+		document.getElementById('addData').addEventListener('click', function() { // TODO : fix issue with addData
```
This seems broken

loicbourgois replied:
See https://github.com/chartjs/Chart.js/issues/5197

benmccann replied:
I would update the comment to link to the issue. It's not clear from the TODO right now what is supposed to be fixed

simonbrunel commented:
It's very hard to review this PR, many modifications that are obfuscated by the indentation change. It might have been better to submit a PR that only fix the indentation without touching the rest of the code.

simonbrunel commented on .htmllintrc.js:
```diff
@@ -0,0 +1,10 @@
+		'indent-style': 'tabs',
+		'attr-quote-style': 'double',
+		'spec-char-escape': false,
+		'attr-bans': ['align', 'background', 'bgcolor', 'border', 'frameborder', 'longdesc', 'marginwidth', 'marginheight', 'scrolling', 'syle'/*, 'width'*/],
```
I think there is a typo to `syle` but do we really need these explicit bans?

loicbourgois replied:
I took the default option and commented the attributes triggering errors.
See https://github.com/htmllint/htmllint/wiki/Options#attr-bans

simonbrunel replied:
I see, then you can remove the 2 last values (`'syle'` which was initially `'style'` and `'width'`)

simonbrunel commented on gulpfile.js:
```diff
@@ -50,8 +52,9 @@ gulp.task('build', buildTask);
 gulp.task('package', packageTask);
 gulp.task('watch', watchTask);
 gulp.task('lint', lintTask);
+gulp.task('lint-html', lintHtmlTask);
```
This task should be ran as a dependency of the `lint` task, not `test`

simonbrunel commented on gulpfile.js:
```diff
@@ -176,6 +193,11 @@ function lintTask() {
 
+function lintHtmlTask() {
+  return gulp.src('./samples/**/*.html')
+    .pipe(htmllint(htmllintOptions));
```
No need to require `htmllintOptions`, I think [this](https://www.npmjs.com/package/gulp-htmllint#optionsconfig) should work:

```javascript
  .pipe(htmllint({
    config: '.htmllint.js' // or .yml is supported
  }));
```

loicbourgois replied:
Looking at the source, gulp-htmllint only supports proper JSON
https://github.com/yvanavermaet/gulp-htmllint/blob/master/src/index.js#L20

simonbrunel commented on gulpfile.js:
```diff
@@ -50,8 +52,9 @@ gulp.task('build', buildTask);
+gulp.task('lint-html', lintHtmlTask);
 gulp.task('docs', docsTask);
-gulp.task('test', ['lint', 'validHTML', 'unittest']);
+gulp.task('test', ['lint', 'lint-html', 'validHTML', 'unittest']);
```
I think htmllint do the same job as `gulp-html-validator` (unmaintained?), so we should remove the `'validHTML'` task and associated node dependencies in package.json.

simonbrunel commented on gulpfile.js:
```diff
@@ -157,7 +160,8 @@ function lintTask() {
     'src/**/*.js',
-    'test/**/*.js'
+    'test/**/*.js',
+    './samples/**/*.html'
```
For consistency:

```javascript
var files = [
    'samples/**/*.html'
    'samples/**/*.js',
    'src/**/*.js',
    'test/**/*.js'
]
```

simonbrunel commented on package.json:
```diff
@@ -17,13 +17,15 @@
     "gulp-connect": "~5.0.0",
     "gulp-eslint": "^4.0.0",
     "gulp-file": "^0.3.0",
     "gulp-html-validator": "^0.0.5",
```
Should be removed (see previous comment)

simonbrunel commented on samples/advanced/data-labelling.html:
```diff
@@ -3,130 +3,130 @@
-        -ms-user-select: none;
-    }
-    </style>
+		<title>Labelling Data Points</title>
```
All files are now indented with 2 tabs instead of 1 for the HTML part

![image](https://user-images.githubusercontent.com/3874900/35537283-f7136e6c-0549-11e8-897d-3111c8d391f6.png)

simonbrunel commented on samples/charts/area/line-boundaries.html:
```diff
@@ -18,14 +18,14 @@
 		<div class="toolbar">
-			<button onclick="toggleSmooth(this)">Smooth</button>
-			<button onclick="randomize(this)">Randomize</button>
+			<button id="toggleSmooth">Smooth</button>
```
What the reason of this change?

loicbourgois replied:
Linting JS in HTML attributes is not supported yet, so I had to move event handling in a `<script>` element.
I submitted a PR to add attributes linting support : https://github.com/BenoitZugmeyer/eslint-plugin-html/pull/87

simonbrunel replied:
I really prefer the other syntax and was about to update other samples to remove these `addEventListener`.

Can we revert these changes?

loicbourgois replied:
If the PR is accepted, yes. 
If not, we have 2 choices : 
* use `onclick()` and add a rule to ignore `no-undef-var` for /samples
* use `addEventListener`

simonbrunel replied:
`no-undef-var: 0` in `samples/.eslintrc.yml` :)

simonbrunel replied:
Alternatively (and probably better), we can disable this rule inline before each method:

```javascript
// eslint-disable-next-line no-undef-var
function randomize() {
//...
```

loicbourgois commented:
@simonbrunel you can add `?w=1` to the url to ignore whitespace in diffs.
see https://github.com/camsong/chrome-github-mate/issues/5
Not sure how well it works. 
I'm happy to proceed with multiple PR if it's easier.

simonbrunel commented:
That's much better with `?w=`, thanks for this tip!

simonbrunel commented:
The only cons with `?w=1`: you can't add review comments

simonbrunel commented on samples/charts/area/line-boundaries.html:
```diff
@@ -18,14 +18,14 @@
 	<script>
 		var presets = window.chartColors;
-		var utils = Samples.utils;
+		var utils = window.Samples.utils;
```
`Samples` needs to be added to the eslintrc globals, then we can remove `window.`

I didn't comment other similar changes, but it needs to be fixed in the whole PR.

loicbourgois commented on gulpfile.js:
```diff
@@ -166,8 +170,21 @@ function lintTask() {
-      'max-statements': [1, 30]
-    }
+      'max-statements': [1, 30],
+      'no-new': 0, // TODO : is this allowed ?
```
I tried `var chart = new Chart(...)` but then it triggers an error because `chart` is not used.

simonbrunel replied:
Makes sense, let's disable `no-new` then

loicbourgois commented:
Something strange is going on.
`htmllint` seems to give different results depending on if  `.htmllintrc.js` or `.htmllintrc.json`is used.
* [commit switching from `.htmllintrc.js` to `.htmllintrc.json`](https://github.com/chartjs/Chart.js/pull/5195/commits/27fa1a7949270ced8c68a4d95f626a6f78e4c064)
* [Travis log with .js](https://travis-ci.org/chartjs/Chart.js/builds/335141924#L2183)
* [Travis log with .json](https://travis-ci.org/chartjs/Chart.js/builds/335144929#L2181)

Also, the build doesn't fail.

benmccann commented on samples/charts/area/line-datasets.html:
```diff
@@ -152,6 +152,18 @@
 			chart.update();
 		}
+
+		document.getElementById('togglePropagate').onclick = function() {
```
Maybe better to use `addEventListener('click'`?

simonbrunel replied:
Better to keep the onclick event handler inline, as DOM attribute (previous code)

benmccann commented on samples/scales/time/line-point-data.html:
```diff
@@ -122,10 +122,13 @@
 			if (config.data.datasets.length > 0) {
-				var numTicks = myLine.scales['x-axis-0'].ticksAsTimestamps.length;
-				var lastTime = numTicks ? moment(myLine.scales['x-axis-0'].ticksAsTimestamps[numTicks - 1]) : moment();
+				// var numTicks = window.myLine.scales['x-axis-0'].ticksAsTimestamps.length; // TODO : check
```
are these commented out because of the issue in referenced in the TODO?

loicbourgois commented:
Fixed indentation for most files (2 tabs -> 1 tab). 
However, I did not find any option to check for this with htmllint. So I added the functionality locally + submitted a [PR](https://github.com/htmllint/htmllint/pull/227). 
This means that this kind of indentation error will not get caught for now.

simonbrunel commented on gulpfile.js:
```diff
@@ -49,9 +49,11 @@ gulp.task('bower', bowerTask);
+gulp.task('lint-html', lintHtmlTask);
+gulp.task('lint-js', lintJsTask);
 gulp.task('docs', docsTask);
-gulp.task('test', ['lint', 'validHTML', 'unittest']);
```
We should also remove the [`validHtml` associated function](/chartjs/Chart.js/blob/master/gulpfile.js#L192)

simonbrunel commented on package.json:
```diff
@@ -17,13 +17,14 @@
     "gulp-eslint": "^4.0.0",
     "gulp-file": "^0.3.0",
-    "gulp-html-validator": "^0.0.5",
+    "gulp-htmllint": "0.0.15",
```
`"^0.0.15"`

simonbrunel commented on samples/charts/area/line-datasets.html:
```diff
@@ -122,7 +122,7 @@
 					propagate: false
 				},
-				samples_filler_analyser: {
+				'samples-filler-analyser': {
```
You also need to change the [plugin id](/chartjs/Chart.js/blob/master/samples/charts/area/analyser.js#L7)

loicbourgois commented:
@simonbrunel what about `no-unused-vars` for inline events ? 
Should we use comments to disable it for now and revert in another PR once https://github.com/BenoitZugmeyer/eslint-plugin-html/pull/87 is merged ?

simonbrunel commented:
You right, it's not `no-undef-var` but `no-unused-vars`. I'm fine with disabling locally in comment (`// eslint-disable-next-line no-unused-var`) or globally at the `samples/.eslintrc.yml`.

loicbourgois commented:
Should be ready for merge
